### PR TITLE
Unit of work

### PR DESCRIPTION
**Pull request by N1NTENDO1999**

### CODE REVIEW

**Review by VoitykRoman — COMMENTED** [review]
l

File: DeratControl.Domain/Root/IUnitOfWork.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+        IFacilityRepository FacilityRepository { get; }
+        IOrganizationRepository OrganizationRepository { get; }
+        IPerimeterRepository PerimeterRepository { get; }
+        IPointRepository PointRepository { get; }
+        IReviewRepository ReviewRepository { get; }
+        ITrapRepository TrapRepository { get; }
+        ITrapReviewRepository TrapReviewRepository { get; }
+        IUserRepository UserRepository { get; }
+
+        void Commit();
```
Comment (SviatoslavMykhailiv) [inline DeratControl.Domain/Root/IUnitOfWork.cs]: make it asynchronous, Task Commit();

File: DeratControl.Ifrastructure/UnitOfWork.cs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        }
+       
+        public DbContext DatabaseContext { get; private set; }
+
+        public UnitOfWork(DbContext dbContext)
+        {
+            DatabaseContext = dbContext;
+        }
+
+        public void Commit()
```
Comment (SviatoslavMykhailiv) [inline DeratControl.Ifrastructure/UnitOfWork.cs]: Task Commit() { return DatabaseContext.SaveChangesAsync();}

File: DeratControl.Ifrastructure/DeratContext.cs [hunk 3]
```diff
@@ -3,21 +3,38 @@
...
 using DeratControl.Infrastructure.EntitiesConfigurations;
+using DeratControl.Domain.Entities;
 
 namespace DeratControl.Infrastructure
 {
-    public class DeratContext:DbContext
+    public class DeratContext : DbContext
     {
+
+        public DbSet<Facility> Facilities { get; set; }
```
Comment (SviatoslavMykhailiv) [inline DeratControl.Ifrastructure/DeratContext.cs]: make all setters for Dbsets private

**Review by SviatoslavMykhailiv — CHANGES_REQUESTED** [review]

File: DeratControl.Ifrastructure/UnitOfWork.cs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        private ITrapReviewRepository trapReviewRepository;
+        private IUserRepository userRepository;
+
+        public IFacilityRepository FacilityRepository
+        {
+            get
+            {
+                if (facilityRepository == null)
+                {
+                    facilityRepository = new FacilityRepository(DatabaseContext);
```
Comment (SviatoslavMykhailiv) [inline DeratControl.Ifrastructure/UnitOfWork.cs:28]: Consider using Lazy<T> class

Comment (SviatoslavMykhailiv) [reply DeratControl.Ifrastructure/UnitOfWork.cs:28]: https://docs.microsoft.com/en-us/dotnet/api/system.lazy-1?view=netframework-4.8

File: DeratControl.Ifrastructure/DeratContext.cs (hunk 3 above)
Comment (N1NTENDO1999) [reply DeratControl.Ifrastructure/DeratContext.cs]: Done

**Review by SviatoslavMykhailiv — APPROVED** [review]

**Pull request merged**